foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_10_0/PomDependencyVersionMigration.java

```
@@ -0,0 +1,147 @@
+            .toList();
+    }
+
+    /**
+     * Returns a list of any dependencies that need to be updated with a new version
+     */
+    private List<Dependency> getMatchingDependenciesWithDifferentVersionsForProject(Model model) {
+        List<Dependency> matchingDependencies = new ArrayList<>();
+
+        this.dependenciesToUpdate.stream()
+            .flatMap(updatedDependency -> 
+                this.getAllDependenciesByFilterForProject(model, dependency -> this.isMatchingDependencyWithDifferentVersion(dependency, updatedDependency))
+                .stream())
+            .forEach(matchingDependencies::add);
+
+        return matchingDependencies;
+    }
+
+    /**
+     * Checks if given dependency matches group id and artifact id but not version of the updated dependency
+     */
+    private boolean isMatchingDependencyWithDifferentVersion(Dependency dependency, Dependency updatedDependency) {
+        return this.isMatchingDependency(dependency, updatedDependency) &&
+               // Return true if the version is not specified (null) or the version is specified but different
+               (dependency.getVersion() == null || !dependency.getVersion().equals(updatedDependency.getVersion()));
+    }
+
+    /**
+     * Checks if two given dependencies match group id and artifact id
+     */
+    private boolean isMatchingDependency(Dependency dependency1, Dependency dependency2) {
+        return dependency1.getGroupId().equals(dependency2.getGroupId()) &&
+               dependency1.getArtifactId().contains(dependency2.getArtifactId());
+    }
+
+    /**
+     * Returns the updated version for the matching dependency
+     */
+    private String getUpdatedDependencyVersion(Dependency dependency) {
+        return this.dependenciesToUpdate.stream()
+            .filter(updatedDependency -> this.isMatchingDependency(updatedDependency, dependency))
+            .map(Dependency::getVersion)
+            .findFirst()
+            .orElse(null);
```

S: Looks like the calling code assumes non-null. Should we use `orElseThrow` instead?

---

Yes that should throw an error if we don't find the version.